// @mantine/code-highlight/cjs/index.cjs
'use strict';

var CodeHighlight = require('./CodeHighlight/CodeHighlight.cjs');
var InlineCodeHighlight = require('./CodeHighlight/InlineCodeHighlight.cjs');
var CodeHighlightTabs = require('./CodeHighlightTabs/CodeHighlightTabs.cjs');
var CodeHighlightControl = require('./CodeHighlight/CodeHighlightControl/CodeHighlightControl.cjs');
var CodeHighlightProvider = require('./CodeHighlightProvider/CodeHighlightProvider.cjs');
var highlightJsAdapter = require('./CodeHighlightProvider/adapters/highlight-js-adapter.cjs');
var shikiAdapter = require('./CodeHighlightProvider/adapters/shiki-adapter.cjs');
var plainTextAdapter = require('./CodeHighlightProvider/adapters/plain-text-adapter.cjs');



exports.CodeHighlight = CodeHighlight.CodeHighlight;
exports.InlineCodeHighlight = InlineCodeHighlight.InlineCodeHighlight;
exports.CodeHighlightTabs = CodeHighlightTabs.CodeHighlightTabs;
exports.CodeHighlightControl = CodeHighlightControl.CodeHighlightControl;
exports.CodeHighlightAdapterProvider = CodeHighlightProvider.CodeHighlightAdapterProvider;
exports.useHighlight = CodeHighlightProvider.useHighlight;
exports.createHighlightJsAdapter = highlightJsAdapter.createHighlightJsAdapter;
exports.createShikiAdapter = shikiAdapter.createShikiAdapter;
exports.stripShikiCodeBlocks = shikiAdapter.stripShikiCodeBlocks;
exports.plainTextAdapter = plainTextAdapter.plainTextAdapter;
